File: day33/houseRobber.cpp
// 198. House Robber
#include<iostream>
#include<vector>
using namespace std;

int maxSumOfNonAdjancentSequence(int i, vector<int> nums){
  if(i==0) return nums[i];
  if(i<0) return 0;

  int pick = nums[i] + maxSumOfNonAdjancentSequence(i-2, nums);
  int notPick = maxSumOfNonAdjancentSequence(i-1, nums);
  return max(pick, notPick);
}

// memoization
int maxSumOfNonAdjancentSequence2(int i, vector<int> nums, vector<int> memo){
  if(i==0) return nums[i];
  if(i<0) return 0;
  if(memo[i] != -1) return memo[i];

  int pick = nums[i] + maxSumOfNonAdjancentSequence2(i-2, nums, memo);
  int notPick = maxSumOfNonAdjancentSequence2(i-1, nums, memo);
  return memo[i] = max(pick, notPick);
}

// tabulation
int maxSumOfNonAdjancentSequence3(vector<int> nums){
  vector<int> dp(nums.size()+1, -1);

  dp[0] = nums[0];
  for(int i=1; i<nums.size(); i++){

    int pick = nums[i] + (i>1? dp[i-2]: 0) ;
    int notPick = 0 + dp[i-1];
    dp[i] = max(pick, notPick);
  }
  return dp[nums.size()-1];
}

// optimized tabulation
int maxSumOfNonAdjancentSequence4(vector<int> nums){
  if(nums.size()== 0 ) return 0;
  if(nums.size()== 1 ) return nums[0]; // Handle single-element case

    int prev2 = 0;  // dp[i-2]
    int prev = nums[0];  // dp[i-1]

  for(int i=1; i<nums.size(); i++){

    int pick = nums[i] + (i>1? prev2: 0) ;
    int notPick = 0 + prev;
    int curr = max(pick, notPick);
    prev2 = prev;
    prev = curr;
  }
  return prev;
}

int main(){
  vector<int> nums = { 2,1,4,9};

  cout<< maxSumOfNonAdjancentSequence(nums.size()-1, nums)<<"\n";

  vector<int> memo (nums.size()+1, -1);
  cout<< maxSumOfNonAdjancentSequence2(nums.size()-1, nums, memo)<<"\n";

  cout<< maxSumOfNonAdjancentSequence3(nums)<<"\n";
  cout<< maxSumOfNonAdjancentSequence4(nums)<<"\n";
  return 0;
}

File: day112/HouseRobber.java
// 198. House Robber
import java.util.Arrays;

public class HouseRobber {
  // private int helper(int[] nums, int i) {
  //   if (i < 0) return 0;
  //   if (i == 0) return nums[0];

  //   int rob = nums[i] + helper(nums, i - 2);
  //   int notRob = helper(nums, i - 1);

  //   return Math.max(rob, notRob);
  // }
  // public int rob(int[] nums) {
  //   return helper(nums, nums.length - 1);
  // }

  private int[] dp;
  // private int helper2(int idx, int[] nums) {
  //   if (idx==0) return nums[idx];
  //   if (idx<0) return 0;
  //   if (dp[idx] != -1) return dp[idx];

  //   // Max(Pick, NotPick)
  //   return dp[idx] = Math.max(nums[idx] + helper2(idx-2, nums), helper2(idx-1, nums));
  // }

  // public int rob2(int[] nums) {
  //   dp = new int[nums.length];
  //   Arrays.fill(dp, -1);
  //   return helper2(nums.length - 1, nums);
  // }

  // public int rob3(int[] nums) {
  //   int n = nums.length;
  //   if(n==0) return 0;
  //   if(n==1) return nums[0];

  //   dp = new int[n];
  //   Arrays.fill(dp, -1);
  //   dp[0] = nums[0];
  //   dp[1] = Math.max(nums[1], nums[0]);
  //   for(int i=2; i<n; ++i){
  //     dp[i] = Math.max(nums[i]+dp[i-2], dp[i-1]);
  //   }

  //   return dp[n-1];
  // }

  public int rob4(int[] nums) {
    int n = nums.length;
    if(n==0) return 0;
    if(n==1) return nums[0];

    int prev2 = nums[0];
    int prev = Math.max(nums[1], nums[0]);
    for(int i=2; i<n; ++i){
      int curr = Math.max(nums[i]+prev2, prev);
      prev2 = prev;
      prev = curr;
    }

    return prev;
  }

  public static void main(String[] args) {
    HouseRobber hr = new HouseRobber();
    System.out.println(hr.rob4(new int[] { 2, 7, 9, 3, 1 }));
  }
}
